File: R/sGARCH.R
#' Implementation of `rugarch` methods for a GARCH(p,q) as R6 class
#'
#' @rdname sGARCH
#' @name sGARCH
#' @keywords GARCH
#' @note Version 1.0.0
#' @export
sGARCH <- R6::R6Class("sGARCH",
                             public = list(
                               #' @description
                               #' Initialize a standard GARCH model
                               #' @param archOrder Integer scalar, ARCH order.
                               #' @param garchOrder Integer scalar, GARCH order.
                               #' @param mode Character, one of \code{"unitOmega"}, \code{"targetSigma2"}, \code{"freeOmega"}.
                               initialize = function(archOrder = 1, garchOrder = 1, mode = "unitOmega"){
                                 # ARCH order
                                 private[["..archOrder"]] <- archOrder
                                 # GARCH order
                                 private[["..garchOrder"]] <- garchOrder
                                 # Unconditional variance constraint
                                 private[["..mode"]] <-  match.arg(mode, choices =  c("unitOmega","targetSigma2","freeOmega"))
                                 # Matrix components
                                 private[["..b"]] <- GARCH_vector_b(archOrder, garchOrder)
                               },
                               #' @description
                               #' Fit the GARCH model with `rugarch` function.
                               #' @param x Numeric, vector. Time series to be fitted.
                               #' @param weights Numeric, vector. Optional custom weights.
                               fit = function(x, weights){
                                 # Time series
                                 private[["..x"]] <- x
                                 # Flexible weights (useful if model do not converge)
                                 if (missing(weights)) {
                                   private[["..w"]] <- rep(1, length(x))
                                 } else {
                                   private[["..w"]] <- ifelse(weights == 0, 0, 1)
                                 }
                                 # GARCH fit
                                 model <- sGARCH_robust_fit(x, private[["..w"]], self$archOrder, self$garchOrder, mode = private[["..mode"]])
                                 # Update coefficients
                                 private[["..omega"]] <- model$omega
                                 private[["..alpha"]] <- model$alpha
                                 private[["..beta"]] <- model$beta
                                 # Log-likelihods
                                 private$..log.likelihoods <- model$loglik
                                 # Std. errors
                                 private[["..std.errors"]] <- model$std.errors
                                 # Matrix components
                                 # Companion matrix
                                 private[["..A"]] <- ARMA_companion_matrix(self$alpha[self$alpha != 0], self$beta[self$beta != 0])
                                 # Intercept vector
                                 private[["..d"]] <-  matrix(c(self$omega, rep(0, sum(self$order)-1)), ncol = 1)
                               },
                               #' @description
                               #' Filter method from `rugarch` package to compute GARCH variance, residuals and log-likelihoods.
                               #' @param x Numeric, vector. Time series to be filtered.
                               #' @param eps0 Optional numeric initial epsilons to prepend (length p+q).
                               #' @param sigma20 Optional numeric initial variances to prepend (length p+q).
                               filter = function(x, eps0 = NULL, sigma20 = NULL){
                                 sGARCH_filter(x, self$omega, self$alpha, self$beta, eps0, sigma20)
                               },
                               #' @description
                               #' Update the coefficients of the model
                               #' @param coefficients Numeric, named vector. Model's coefficients.
                               update = function(coefficients){
                                 if (missing(coefficients)) {
                                   return(invisible(NULL))
                                 }
                                 # Extract old coefficients
                                 new_coefs <- self$coefficients
                                 # Extract names
                                 names_old <- names(new_coefs)
                                 names_new <- names(coefficients)
                                 # Warning
                                 if (length(names_new) != length(names_old)) {
                                   cli::cli_alert_warning("In sGARCH$update(): The lenght of new `coefficients` do not match the length of the old coefficients.")
                                 }
                                 # Update only if they are present
                                 for(i in 1:length(coefficients)){
                                   if (names_new[i] %in% names_old) {
                                     new_coefs[names_new[i]] <- coefficients[i]
                                     private$..std.errors[names_new[i]] <- NA_integer_
                                   }
                                 }
                                 # Update intercept
                                 private$..omega <- new_coefs["omega"]
                                 # Update AR parameters
                                 if (self$archOrder > 0) {
                                   private$..alpha <- new_coefs[stringr::str_detect(names_old, "alpha")]
                                 }
                                 # Update MA parameters
                                 if (self$garchOrder > 0) {
                                   private$..beta <- new_coefs[stringr::str_detect(names_old, "beta")]
                                 }
                                 # Matrix components
                                 # Companion matrix
                                 private[["..A"]] <- ARMA_companion_matrix(self$alpha[self$alpha != 0], self$beta[self$beta != 0])
                                 # Intercept vector
                                 private[["..d"]] <-  matrix(c(self$omega, rep(0, sum(self$order)-1)), ncol = 1)
                               },
                               #' @description
                               #' Numerical computation of the std. errors of the parameters.
                               #' @param std.errors Numeric std. errors.
                               update_std.errors = function(std.errors){
                                 if (missing(std.errors) || purrr::is_empty(std.errors)) {
                                   return(invisible(NULL))
                                 }
                                 # Extract old coefficients
                                 new_std.errors <- private$..std.errors
                                 # Extract names
                                 names_old <- names(new_std.errors)
                                 names_new <- names(std.errors)
                                 # Warning
                                 if (length(names_new) != length(names_old)) {
                                   cli::cli_alert_warning("In sGARCH$update_std.errors(): The lenght of new `std.errors` do not match the length of the old std. errors!")
                                 }
                                 # Update only if they are present
                                 for(i in 1:length(std.errors)){
                                   if (names_new[i] %in% names_old) {
                                     new_std.errors[names_new[i]] <- std.errors[i]
                                   }
                                 }
                                 # Update private std. errors
                                 private$..std.errors <- new_std.errors
                               },
                               #' @description
                               #' Next step GARCH std. deviation forecast
                               #' @param eps0 Numeric initial epsilons to prepend (length p+q).
                               #' @param sigma20 Numeric initial variances to prepend (length p+q).
                               next_step = function(eps0, sigma20){
                                 sGARCH_next_step(self$omega, self$alpha, self$beta, eps0, sigma20)
                               },
                               #' @description
                               #' Print method for `GARCH_modelR6` class.
                               print = function(){
                                 # Format parameters with std. errors
                                 green <- function(x) paste0("\033[1;32m", x, "\033[0m")
                                 red <- function(x) paste0("\033[1;31m", x, "\033[0m")
                                 msg_col <- function(x) ifelse(x, green(x), red(x))
                                 # Intercept
                                 par.intercept <- format(self$omega, digits = 4)
                                 std.error.intercept <- self$std.errors[names(self$std.errors) == names(par.intercept)]
                                 format.intercept <- paste0(par.intercept, " (", format(std.error.intercept, digits = 3), ")")
                                 # ARCH parameters
                                 archOrder <- self$archOrder
                                 par.arch <- format(self$alpha, digits = 4)
                                 std.error.arch <- self$std.errors[names(self$std.errors) %in% names(par.arch)]
                                 format.arch <- paste0(paste0(par.arch, " (", format(std.error.arch, digits = 3), ")"), collapse = " ")
                                 # GARCH parameters
                                 garchOrder <- self$garchOrder
                                 par.garch <- format(self$beta, digits = 4)
                                 std.error.garch <- self$std.errors[names(self$std.errors) %in% names(par.garch)]
                                 format.garch <- paste0(paste0(par.garch, " (", format(std.error.garch, digits = 3), ")"), collapse = " ")
                                 # Model name
                                 model_name <- paste0("GARCH", " (", self$order[1], ", ", self$order[2], ")")
                                 # ********************************************************
                                 msg_0 <- paste0("--------------------- ", model_name, "--------------------- \n")
                                 msg_1 <- paste0("ARCH: ", msg_col(archOrder != 0), "\n")
                                 msg_2 <- paste0("GARCH: ", msg_col(garchOrder != 0), "\n")
                                 msg_3 <- paste0("Version: ", private$version, "\n")
                                 msg_4 <- paste0("Unconditional Variance: ", private$..sigma20, "\n")
                                 msg_5 <- "------------------------------------------------\n"
                                 msg_6 <- paste0("Intercept: ", format.intercept, " \n")
                                 msg_7 <- paste0("Alpha: ", format.arch, " \n")
                                 msg_8 <- paste0("Beta: ", format.garch, " \n")
                                 msg_9 <- paste0("Log-lik: ", format(self$loglik, digits = 3), "\n")
                                 cat(c(msg_0, msg_1, msg_2, msg_3, msg_4,
                                       msg_5, msg_6, msg_7, msg_8, msg_9))
                               }
                             ),
                             private = list(
                               version = "1.0.1",
                               ..archOrder = 0,
                               ..garchOrder = 0,
                               ..x = NA,
                               ..w = NA,
                               ..mode = NA,
                               ..omega = c(omega = 1),
                               ..alpha = c(alpha1 = 0),
                               ..beta = c(beta1 = 0),
                               ..A = matrix(),
                               ..b = matrix(),
                               ..d = matrix(),
                               ..log.likelihoods = c(),
                               ..std.errors = c(NA, NA, NA)
                             ),
                             active = list(
                               #' @field archOrder Numeric scalar, ARCH order.
                               archOrder = function(){
                                 private$..archOrder
                               },
                               #' @field garchOrder Numeric scalar, GARCH order.
                               garchOrder = function(){
                                 private$..garchOrder
                               },
                               #' @field order Numeric named vector, orders of the GARCH model. The first element is the ARCH order, while the second the GARCH order.
                               order = function(){
                                 c(ARCH = self$archOrder, GARCH = self$garchOrder)
                               },
                               #' @field omega Numeric scalar, intercept parameter.
                               omega = function(){
                                 private$..omega
                               },
                               #' @field alpha Numeric vector, ARCH parameters.
                               alpha = function(){
                                 private$..alpha
                               },
                               #' @field beta Numeric vector, GARCH parameters.
                               beta = function(){
                                 private$..beta
                               },
                               #' @field coefficients Numeric vector, model's coefficients.
                               coefficients = function(){
                                 c(self$omega, self$alpha, self$beta)
                               },
                               #' @field A Numeric matrix, companion matrix.
                               A = function(){
                                 private$..A
                               },
                               #' @field b Numeric vector.
                               b = function(){
                                 private$..b
                               },
                               #' @field d Numeric vector
                               d = function(){
                                 private$..d
                               },
                               #' @field std.errors Numeric vector, standard errors of the model's coefficients.
                               std.errors = function(){
                                 private$..std.errors
                               },
                               #' @field sigma2_inf Numeric scalar, long-term unconditional std. deviation.
                               sigma2_inf = function(){
                                 self$omega / (1 - sum(self$alpha+self$beta))
                               },
                               #' @field loglik model log-likelihood
                               loglik = function(){
                                 sum(private$..log.likelihoods)
                               },
                               #' @field tidy Tibble with estimated parameters and relative std. errors.
                               tidy = function(){
                                 # Extract the parameters
                                 params <- self$coefficients
                                 dplyr::tibble(
                                   term = names(params),
                                   estimate = params,
                                   std.error = self$std.errors,
                                 )
                               }
                             )
)

File: solarrpy/sGARCH.py
import numpy as np
import pandas as pd
from typing import Optional, Tuple, Dict, List
import warnings
from arch import arch_model


class sGARCH:
    """
    Implementation of standard GARCH(p,q) model as a Python class.
    
    This class provides methods to fit, filter, and forecast GARCH models,
    similar to the rugarch package in R.
    
    Example:
    --------
    # Initialize GARCH(1,1) model
    model = sGARCH(arch_order=1, garch_order=1)
    
    # Fit the model
    model.fit(x=returns)
    
    # Filter to get conditional variance
    sigma2 = model.filter(x=returns)
    
    # Forecast next step
    next_var = model.next_step(eps0=[0.01], sigma20=[0.005])
    
    Version: 1.0.1
    Keywords: GARCH
    """
    
    VERSION = "1.0.1"
    
    def __init__(self, arch_order: int = 1, garch_order: int = 1, 
                 mode: str = "unitOmega"):
        """
        Initialize a standard GARCH model.
        
        Parameters:
        -----------
        arch_order : int
            ARCH order (p)
        garch_order : int
            GARCH order (q)
        mode : str
            Unconditional variance constraint: 'unitOmega', 'targetSigma2', or 'freeOmega'
        """
        # ARCH order
        self._arch_order = arch_order
        
        # GARCH order
        self._garch_order = garch_order
        
        # Unconditional variance constraint
        if mode not in ['unitOmega', 'targetSigma2', 'freeOmega']:
            raise ValueError("mode must be one of: 'unitOmega', 'targetSigma2', 'freeOmega'")
        self._mode = mode
        
        # Initialize parameters
        self._omega = np.array([1.0])
        self._alpha = np.zeros(arch_order) if arch_order > 0 else np.array([0.0])
        self._beta = np.zeros(garch_order) if garch_order > 0 else np.array([0.0])
        
        # Matrix components
        self._b = self._GARCH_vector_b(arch_order, garch_order)
        self._A = np.array([[]])
        self._d = np.array([[]])
        
        # Data storage
        self._x = None
        self._w = None
        self._log_likelihoods = np.array([])
        self._std_errors = np.array([np.nan, np.nan, np.nan])
        self._sigma20 = None
    
    @staticmethod
    def _GARCH_vector_b(arch_order: int, garch_order: int) -> np.ndarray:
        """
        Create the b vector for GARCH companion form.
        
        Parameters:
        -----------
        arch_order : int
            ARCH order
        garch_order : int
            GARCH order
        
        Returns:
        --------
        np.ndarray
            Vector b
        """
        max_order = max(arch_order, garch_order)
        b = np.zeros((max_order, 1))
        b[0, 0] = 1.0
        return b
    
    @staticmethod
    def _ARMA_companion_matrix(phi: np.ndarray, theta: np.ndarray) -> np.ndarray:
        """
        Create ARMA companion matrix.
        
        Parameters:
        -----------
        phi : np.ndarray
            AR coefficients
        theta : np.ndarray
            MA coefficients
        
        Returns:
        --------
        np.ndarray
            Companion matrix
        """
        p = len(phi)
        q = len(theta)
        m = max(p, q)
        
        if m == 0:
            return np.array([[]])
        
        # Create companion matrix
        A = np.zeros((m, m))
        
        # First row contains AR and MA coefficients
        if p > 0:
            A[0, :p] = phi
        if q > 0:
            A[0, :q] += theta
        
        # Identity submatrix
        if m > 1:
            A[1:, :-1] = np.eye(m - 1)
        
        return A
    
    def fit(self, x: np.ndarray, weights: Optional[np.ndarray] = None):
        """
        Fit the GARCH model.
        
        Parameters:
        -----------
        x : np.ndarray
            Time series to be fitted
        weights : np.ndarray, optional
            Custom weights for observations
        """
        # Store time series
        self._x = np.array(x)
        
        # Flexible weights
        if weights is None:
            self._w = np.ones(len(x))
        else:
            self._w = np.where(weights == 0, 0, 1)
        
        # Fit using arch package
        try:
            # Create GARCH specification
            am = arch_model(
                x,
                vol='GARCH',
                p=self._arch_order,
                q=self._garch_order,
                mean='Zero',
                rescale=False
            )
            
            # Fit the model
            res = am.fit(disp='off')
            
            # Extract parameters
            params = res.params
            
            # Update coefficients
            self._omega = np.array([params['omega']])
            
            if self._arch_order > 0:
                self._alpha = np.array([params[f'alpha[{i+1}]'] 
                                       for i in range(self._arch_order)])
            
            if self._garch_order > 0:
                self._beta = np.array([params[f'beta[{i+1}]'] 
                                      for i in range(self._garch_order)])
            
            # Log-likelihood
            self._log_likelihoods = -res.resid**2 / (2 * res.conditional_volatility**2) - \
                                   0.5 * np.log(2 * np.pi * res.conditional_volatility**2)
            
            # Standard errors
            std_errs = []
            std_errs.append(res.std_err['omega'])
            for i in range(self._arch_order):
                std_errs.append(res.std_err[f'alpha[{i+1}]'])
            for i in range(self._garch_order):
                std_errs.append(res.std_err[f'beta[{i+1}]'])
            self._std_errors = np.array(std_errs)
            
            # Unconditional variance
            self._sigma20 = res.conditional_volatility.mean()**2
            
        except Exception as e:
            warnings.warn(f"GARCH fitting failed: {e}. Using fallback method.")
            self._fallback_fit(x, weights)
        
        # Update matrix components
        alpha_nonzero = self._alpha[self._alpha != 0]
        beta_nonzero = self._beta[self._beta != 0]
        
        if len(alpha_nonzero) > 0 or len(beta_nonzero) > 0:
            self._A = self._ARMA_companion_matrix(alpha_nonzero, beta_nonzero)
        
        # Intercept vector
        total_order = self._arch_order + self._garch_order
        if total_order > 0:
            self._d = np.zeros((total_order, 1))
            self._d[0, 0] = self._omega[0]
    
    def _fallback_fit(self, x: np.ndarray, weights: Optional[np.ndarray]):
        """Fallback fitting method using simple moment estimation."""
        # Simple moment-based estimation
        self._omega = np.array([np.var(x) * 0.1])
        
        if self._arch_order > 0:
            self._alpha = np.full(self._arch_order, 0.1)
        
        if self._garch_order > 0:
            self._beta = np.full(self._garch_order, 0.8)
        
        self._std_errors = np.full(1 + self._arch_order + self._garch_order, np.nan)
        self._sigma20 = np.var(x)
    
    def filter(self, x: np.ndarray, eps0: Optional[np.ndarray] = None,
               sigma20: Optional[np.ndarray] = None) -> np.ndarray:
        """
        Filter method to compute GARCH variance.
        
        Parameters:
        -----------
        x : np.ndarray
            Time series to be filtered
        eps0 : np.ndarray, optional
            Initial epsilons (length p+q)
        sigma20 : np.ndarray, optional
            Initial variances (length p+q)
        
        Returns:
        --------
        np.ndarray
            Conditional variances
        """
        return self._sGARCH_filter(x, self._omega[0], self._alpha, self._beta, 
                                   eps0, sigma20)
    
    def _sGARCH_filter(self, x: np.ndarray, omega: float, 
                       alpha: np.ndarray, beta: np.ndarray,
                       eps0: Optional[np.ndarray] = None,
                       sigma20: Optional[np.ndarray] = None) -> np.ndarray:
        """
        Compute GARCH filtered variances.
        
        Parameters:
        -----------
        x : np.ndarray
            Time series
        omega : float
            Intercept
        alpha : np.ndarray
            ARCH coefficients
        beta : np.ndarray
            GARCH coefficients
        eps0 : np.ndarray, optional
            Initial residuals
        sigma20 : np.ndarray, optional
            Initial variances
        
        Returns:
        --------
        np.ndarray
            Conditional variances
        """
        n = len(x)
        p = len(alpha)
        q = len(beta)
        
        # Initialize
        if eps0 is None:
            eps0 = np.zeros(p)
        if sigma20 is None:
            sigma20 = np.full(q, omega / (1 - np.sum(alpha) - np.sum(beta)))
        
        # Storage
        sigma2 = np.zeros(n)
        eps = np.concatenate([eps0, x])
        
        # Recursion
        for t in range(n):
            # ARCH component
            arch_sum = np.sum(alpha * eps[t:t+p][::-1]**2) if p > 0 else 0
            
            # GARCH component
            if t < q:
                available_sigma2 = np.concatenate([sigma20[:q-t][::-1], sigma2[:t][::-1]])
            else:
                available_sigma2 = sigma2[t-q:t][::-1]
            garch_sum = np.sum(beta * available_sigma2) if q > 0 else 0
            
            # Update variance
            sigma2[t] = omega + arch_sum + garch_sum
        
        return sigma2
    
    def update(self, coefficients: Optional[Dict[str, float]] = None):
        """
        Update the coefficients of the model.
        
        Parameters:
        -----------
        coefficients : dict, optional
            Dictionary of coefficient names and values
        """
        if coefficients is None:
            return
        
        # Get current coefficients
        current_coefs = self.coefficients
        
        # Update omega
        if 'omega' in coefficients:
            self._omega = np.array([coefficients['omega']])
            if 'omega' in range(len(self._std_errors)):
                self._std_errors[0] = np.nan
        
        # Update alpha parameters
        for i in range(self._arch_order):
            key = f'alpha{i+1}'
            if key in coefficients:
                self._alpha[i] = coefficients[key]
                if i + 1 < len(self._std_errors):
                    self._std_errors[i + 1] = np.nan
        
        # Update beta parameters
        for i in range(self._garch_order):
            key = f'beta{i+1}'
            if key in coefficients:
                self._beta[i] = coefficients[key]
                if i + 1 + self._arch_order < len(self._std_errors):
                    self._std_errors[i + 1 + self._arch_order] = np.nan
        
        # Update matrix components
        alpha_nonzero = self._alpha[self._alpha != 0]
        beta_nonzero = self._beta[self._beta != 0]
        
        if len(alpha_nonzero) > 0 or len(beta_nonzero) > 0:
            self._A = self._ARMA_companion_matrix(alpha_nonzero, beta_nonzero)
        
        # Update intercept vector
        total_order = sum(self.order)
        if total_order > 0:
            self._d = np.zeros((total_order, 1))
            self._d[0, 0] = self._omega[0]
    
    def update_std_errors(self, std_errors: Optional[Dict[str, float]] = None):
        """
        Update the standard errors of the parameters.
        
        Parameters:
        -----------
        std_errors : dict, optional
            Dictionary of standard errors
        """
        if std_errors is None or len(std_errors) == 0:
            return
        
        # Create new std errors array
        new_std_errors = self._std_errors.copy()
        
        # Map of parameter names to indices
        idx = 0
        param_map = {'omega': idx}
        idx += 1
        
        for i in range(self._arch_order):
            param_map[f'alpha{i+1}'] = idx
            idx += 1
        
        for i in range(self._garch_order):
            param_map[f'beta{i+1}'] = idx
            idx += 1
        
        # Update values
        for key, value in std_errors.items():
            if key in param_map:
                new_std_errors[param_map[key]] = value
        
        self._std_errors = new_std_errors
    
    def next_step(self, eps0: np.ndarray, sigma20: np.ndarray) -> float:
        """
        Compute next step GARCH variance forecast.
        
        Parameters:
        -----------
        eps0 : np.ndarray
            Recent residuals (length p)
        sigma20 : np.ndarray
            Recent variances (length q)
        
        Returns:
        --------
        float
            Next period variance forecast
        """
        # ARCH component
        arch_sum = np.sum(self._alpha * np.array(eps0)**2) if self._arch_order > 0 else 0
        
        # GARCH component
        garch_sum = np.sum(self._beta * np.array(sigma20)) if self._garch_order > 0 else 0
        
        # Next variance
        return self._omega[0] + arch_sum + garch_sum
    
    def __repr__(self) -> str:
        """Print method for sGARCH class."""
        # Format parameters with std. errors
        def format_param(val, stderr):
            return f"{val:.4f} ({stderr:.3f})" if not np.isnan(stderr) else f"{val:.4f} (NA)"
        
        # Intercept
        omega_str = format_param(self._omega[0], self._std_errors[0])
        
        # ARCH parameters
        arch_strs = []
        for i in range(self._arch_order):
            idx = 1 + i
            arch_strs.append(format_param(self._alpha[i], 
                                         self._std_errors[idx] if idx < len(self._std_errors) else np.nan))
        alpha_str = " ".join(arch_strs) if arch_strs else "None"
        
        # GARCH parameters
        garch_strs = []
        for i in range(self._garch_order):
            idx = 1 + self._arch_order + i
            garch_strs.append(format_param(self._beta[i],
                                          self._std_errors[idx] if idx < len(self._std_errors) else np.nan))
        beta_str = " ".join(garch_strs) if garch_strs else "None"
        
        # Model name
        model_name = f"GARCH({self._arch_order}, {self._garch_order})"
        
        lines = [
            f"--------------------- {model_name} ---------------------",
            f"ARCH: {self._arch_order > 0}",
            f"GARCH: {self._garch_order > 0}",
            f"Version: {self.VERSION}",
            f"Unconditional Variance: {self._sigma20}",
            "------------------------------------------------",
            f"Intercept: {omega_str}",
            f"Alpha: {alpha_str}",
            f"Beta: {beta_str}",
            f"Log-lik: {self.loglik:.3f}" if self.loglik is not None else "Log-lik: NA"
        ]
        
        return "\n".join(lines)
    
    # ==================== Properties ====================
    
    @property
    def arch_order(self) -> int:
        """ARCH order."""
        return self._arch_order
    
    @property
    def garch_order(self) -> int:
        """GARCH order."""
        return self._garch_order
    
    @property
    def order(self) -> Tuple[int, int]:
        """Model orders (ARCH, GARCH)."""
        return (self._arch_order, self._garch_order)
    
    @property
    def omega(self) -> np.ndarray:
        """Intercept parameter."""
        return self._omega
    
    @property
    def alpha(self) -> np.ndarray:
        """ARCH parameters."""
        return self._alpha
    
    @property
    def beta(self) -> np.ndarray:
        """GARCH parameters."""
        return self._beta
    
    @property
    def coefficients(self) -> Dict[str, float]:
        """Model coefficients as dictionary."""
        coefs = {'omega': self._omega[0]}
        
        for i in range(self._arch_order):
            coefs[f'alpha{i+1}'] = self._alpha[i]
        
        for i in range(self._garch_order):
            coefs[f'beta{i+1}'] = self._beta[i]
        
        return coefs
    
    @property
    def A(self) -> np.ndarray:
        """Companion matrix."""
        return self._A
    
    @property
    def b(self) -> np.ndarray:
        """Vector b."""
        return self._b
    
    @property
    def d(self) -> np.ndarray:
        """Intercept vector."""
        return self._d
    
    @property
    def std_errors(self) -> np.ndarray:
        """Standard errors of coefficients."""
        return self._std_errors
    
    @property
    def sigma2_inf(self) -> float:
        """Long-term unconditional variance."""
        denom = 1 - np.sum(self._alpha) - np.sum(self._beta)
        if denom <= 0:
            warnings.warn("Model is not stationary (sum of alpha + beta >= 1)")
            return np.inf
        return self._omega[0] / denom
    
    @property
    def loglik(self) -> Optional[float]:
        """Model log-likelihood."""
        if len(self._log_likelihoods) == 0:
            return None
        return np.sum(self._log_likelihoods)
    
    @property
    def tidy(self) -> pd.DataFrame:
        """Tibble with estimated parameters and std. errors."""
        coefs = self.coefficients
        
        terms = list(coefs.keys())
        estimates = list(coefs.values())
        std_errs = list(self._std_errors[:len(terms)])
        
        return pd.DataFrame({
            'term': terms,
            'estimate': estimates,
            'std_error': std_errs
        })


# Example usage
if __name__ == "__main__":
    # Simulate some returns data
    np.random.seed(42)
    n = 1000
    returns = np.random.randn(n) * 0.01
    
    # Create and fit GARCH(1,1) model
    model = sGARCH(arch_order=1, garch_order=1)
    model.fit(returns)
    
    print(model)
    print("\nTidy output:")
    print(model.tidy)
    
    # Filter to get conditional variance
    sigma2 = model.filter(returns)
    print(f"\nMean conditional variance: {np.mean(sigma2):.6f}")
    
    # Next step forecast
    next_var = model.next_step(eps0=[returns[-1]], sigma20=[sigma2[-1]])
    print(f"Next period variance forecast: {next_var:.6f}")
